File: internal/engine/c_api/api_data/table.cc
/**
 * Copyright 2019 The Gamma Authors.
 *
 * This source code is licensed under the Apache License, Version 2.0 license
 * found in the LICENSE file in the root directory of this source tree.
 */

#include "table.h"

#include "util/utils.h"

namespace vearch {

int TableInfo::Serialize(char **out, int *out_len) {
  flatbuffers::FlatBufferBuilder builder;

  std::vector<flatbuffers::Offset<gamma_api::FieldInfo>> field_info_vector;
  for (const struct FieldInfo &f : fields_) {
    auto field = gamma_api::CreateFieldInfo(
        builder, builder.CreateString(f.name),
        static_cast<::DataType>(f.data_type), f.is_index);
    field_info_vector.push_back(field);
  }

  std::vector<flatbuffers::Offset<gamma_api::VectorInfo>> vector_info_vector;
  for (const struct VectorInfo &v : vectors_infos_) {
    auto vectorinfo = gamma_api::CreateVectorInfo(
        builder, builder.CreateString(v.name),
        static_cast<::DataType>(v.data_type), v.is_index, v.dimension,
        builder.CreateString(v.store_type),
        builder.CreateString(v.store_param));
    vector_info_vector.push_back(vectorinfo);
  }

  auto table = gamma_api::CreateTable(builder, builder.CreateString(name_),
                                      builder.CreateVector(field_info_vector),
                                      builder.CreateVector(vector_info_vector),
                                      compress_mode_,
                                      builder.CreateString(index_type_),
                                      builder.CreateString(index_params_));
  builder.Finish(table);
  *out_len = builder.GetSize();
  *out = (char *)malloc(*out_len * sizeof(char));
  memcpy(*out, (char *)builder.GetBufferPointer(), *out_len);

  return 0;
}

void TableInfo::Deserialize(const char *data, int len) {
  table_ = const_cast<gamma_api::Table *>(gamma_api::GetTable(data));

  name_ = table_->name()->str();

  for (size_t i = 0; i < table_->fields()->size(); ++i) {
    auto f = table_->fields()->Get(i);
    struct FieldInfo field_info;
    field_info.name = f->name()->str();
    field_info.data_type = static_cast<DataType>(f->data_type());
    field_info.is_index = f->is_index();

    fields_.emplace_back(field_info);
  }

  for (size_t i = 0; i < table_->vectors_info()->size(); ++i) {
    auto v = table_->vectors_info()->Get(i);

    struct VectorInfo vector_info;
    vector_info.name = v->name()->str();
    vector_info.data_type = static_cast<DataType>(v->data_type());
    vector_info.is_index = v->is_index();
    vector_info.dimension = v->dimension();
    vector_info.store_type = v->store_type()->str();
    vector_info.store_param = v->store_param()->str();

    vectors_infos_.emplace_back(vector_info);
  }

  index_type_ = table_->index_type()->str();
  index_params_ = table_->index_params()->str();
  utils::JsonParser params_parser;
  int training_threshold = 0;
  params_parser.GetInt("training_threshold", training_threshold);
  if (training_threshold > 0) {
    training_threshold_ = training_threshold;
  }
  compress_mode_ = table_->compress_mode();
}

std::string &TableInfo::Name() { return name_; }

void TableInfo::SetName(std::string &name) { name_ = name; }

std::vector<struct FieldInfo> &TableInfo::Fields() { return fields_; }

std::vector<struct VectorInfo> &TableInfo::VectorInfos() {
  return vectors_infos_;
}

void TableInfo::AddVectorInfo(struct VectorInfo &vector_info) {
  vectors_infos_.emplace_back(vector_info);
}

void TableInfo::AddField(struct FieldInfo &field) {
  fields_.emplace_back(field);
}

int TableInfo::TrainingThreshold() { return training_threshold_; }

void TableInfo::SetTrainingThreshold(int training_threshold) {
  training_threshold_ = training_threshold;
}

std::string &TableInfo::IndexType() { return index_type_; }

void TableInfo::SetIndexType(std::string &index_type) {
  index_type_ = index_type;
}

std::string &TableInfo::IndexParams() { return index_params_; }

void TableInfo::SetIndexParams(std::string &index_params) {
  index_params_ = index_params;
}

int TableInfo::Read(const std::string &path) {
  assert(table_ == nullptr);
  long file_size = utils::get_file_size(path);
  FILE *fp = fopen(path.c_str(), "rb");

  char data[file_size];
  fread(data, file_size, 1, fp);
  fclose(fp);

  Deserialize(data, file_size);
  return 0;
}

int TableInfo::Write(const std::string &path) {
  char *data;
  int len;
  Serialize(&data, &len);

  FILE *fp = fopen(path.c_str(), "wb");

  fwrite(data, len, 1, fp);
  fclose(fp);
  free(data);
  return 0;
}
}  // namespace vearch

File: internal/engine/sdk/go/gamma/table.go
/**
 * Copyright 2019 The Vearch Authors.
 *
 * This source code is licensed under the Apache License, Version 2.0 license
 * found in the LICENSE file in the root directory of this source tree.
 */

package gamma

import (
	flatbuffers "github.com/google/flatbuffers/go"
	"github.com/vearch/vearch/v3/internal/engine/idl/fbs-gen/go/gamma_api"
)

type DataType int8

const (
	INT         DataType = 0
	LONG        DataType = 1
	FLOAT       DataType = 2
	DOUBLE      DataType = 3
	STRING      DataType = 4
	VECTOR      DataType = 5
	BOOL        DataType = 6
	DATE        DataType = 7
	STRINGARRAY DataType = 8
)

type VectorInfo struct {
	Name       string
	DataType   DataType
	IsIndex    bool
	Dimension  int32
	StoreType  string
	StoreParam string
}

type FieldInfo struct {
	Name      string
	DataType  DataType
	IsIndex   bool
	IndexType int32
}

type IndexInfo struct {
	Name       string
	Type       string
	FieldName  string
	FieldNames []string
	Params     string
}

type Table struct {
	Name            string
	Fields          []FieldInfo
	VectorsInfos    []VectorInfo
	IndexType       string
	IndexParams     string
	RefreshInterval int32
	EnableIdCache   bool
	EnableRealtime  bool
	Indexes         []IndexInfo
	table           *gamma_api.Table
}

func (table *Table) Serialize() []byte {
	builder := flatbuffers.NewBuilder(0)
	name := builder.CreateString(table.Name)

	fieldNames := make([]flatbuffers.UOffsetT, len(table.Fields))
	for i := 0; i < len(table.Fields); i++ {
		field := table.Fields[i]
		fieldNames[i] = builder.CreateString(field.Name)
	}
	fieldInfos := make([]flatbuffers.UOffsetT, len(table.Fields))
	for i := 0; i < len(table.Fields); i++ {
		field := table.Fields[i]
		indexType := field.IndexType
		if indexType == 0 && field.IsIndex {
			indexType = 1
		}
		gamma_api.FieldInfoStart(builder)
		gamma_api.FieldInfoAddName(builder, fieldNames[i])
		gamma_api.FieldInfoAddDataType(builder, int8(field.DataType))
		gamma_api.FieldInfoAddIsIndex(builder, indexType != 0)
		gamma_api.FieldInfoAddIndexType(builder, indexType)
		fieldInfos[i] = gamma_api.FieldInfoEnd(builder)
	}

	gamma_api.TableStartFieldsVector(builder, len(table.Fields))
	for i := len(table.Fields) - 1; i >= 0; i-- {
		builder.PrependUOffsetT(fieldInfos[i])
	}
	fields := builder.EndVector(len(table.Fields))

	var names, storeTypes, storeParams []flatbuffers.UOffsetT
	names = make([]flatbuffers.UOffsetT, len(table.VectorsInfos))
	storeTypes = make([]flatbuffers.UOffsetT, len(table.VectorsInfos))
	storeParams = make([]flatbuffers.UOffsetT, len(table.VectorsInfos))
	for i := 0; i < len(table.VectorsInfos); i++ {
		vecInfo := table.VectorsInfos[i]
		names[i] = builder.CreateString(vecInfo.Name)
		storeTypes[i] = builder.CreateString(vecInfo.StoreType)
		storeParams[i] = builder.CreateString(vecInfo.StoreParam)
	}

	vectorInfos := make([]flatbuffers.UOffsetT, len(table.VectorsInfos))
	for i := 0; i < len(table.VectorsInfos); i++ {
		vecInfo := table.VectorsInfos[i]
		gamma_api.VectorInfoStart(builder)
		gamma_api.VectorInfoAddName(builder, names[i])
		gamma_api.VectorInfoAddDataType(builder, int8(vecInfo.DataType))
		gamma_api.VectorInfoAddIsIndex(builder, vecInfo.IsIndex)
		gamma_api.VectorInfoAddDimension(builder, vecInfo.Dimension)
		gamma_api.VectorInfoAddStoreType(builder, storeTypes[i])
		gamma_api.VectorInfoAddStoreParam(builder, storeParams[i])
		vectorInfos[i] = gamma_api.VectorInfoEnd(builder)
	}

	gamma_api.TableStartVectorsInfoVector(builder, len(table.VectorsInfos))
	for i := len(table.VectorsInfos) - 1; i >= 0; i-- {
		builder.PrependUOffsetT(vectorInfos[i])
	}
	vecInfos := builder.EndVector(len(table.VectorsInfos))

	var indexesOffsets []flatbuffers.UOffsetT
	for _, idx := range table.Indexes {
		name := builder.CreateString(idx.Name)
		idxType := builder.CreateString(idx.Type)
		fieldName := builder.CreateString(idx.FieldName)
		params := builder.CreateString(idx.Params)

		fieldNameOffsets := make([]flatbuffers.UOffsetT, len(idx.FieldNames))
		for j := 0; j < len(idx.FieldNames); j++ {
			fieldNameOffsets[j] = builder.CreateString(idx.FieldNames[j])
		}
		gamma_api.IndexInfoStartFieldNamesVector(builder, len(idx.FieldNames))
		for j := len(idx.FieldNames) - 1; j >= 0; j-- {
			builder.PrependUOffsetT(fieldNameOffsets[j])
		}
		fieldNamesVec := builder.EndVector(len(idx.FieldNames))

		gamma_api.IndexInfoStart(builder)
		gamma_api.IndexInfoAddName(builder, name)
		gamma_api.IndexInfoAddType(builder, idxType)
		gamma_api.IndexInfoAddFieldName(builder, fieldName)
		gamma_api.IndexInfoAddFieldNames(builder, fieldNamesVec)
		gamma_api.IndexInfoAddParams(builder, params)
		indexesOffsets = append(indexesOffsets, gamma_api.IndexInfoEnd(builder))
	}

	var indexesVec flatbuffers.UOffsetT
	if len(indexesOffsets) > 0 {
		gamma_api.TableStartIndexesVector(builder, len(indexesOffsets))
		for i := len(indexesOffsets) - 1; i >= 0; i-- {
			builder.PrependUOffsetT(indexesOffsets[i])
		}
		indexesVec = builder.EndVector(len(indexesOffsets))
	}

	indexType := builder.CreateString(table.IndexType)
	indexParams := builder.CreateString(table.IndexParams)

	gamma_api.TableStart(builder)
	gamma_api.TableAddName(builder, name)
	gamma_api.TableAddFields(builder, fields)
	gamma_api.TableAddVectorsInfo(builder, vecInfos)
	gamma_api.TableAddIndexType(builder, indexType)
	gamma_api.TableAddIndexParams(builder, indexParams)
	gamma_api.TableAddRefreshInterval(builder, table.RefreshInterval)
	gamma_api.TableAddEnableIdCache(builder, table.EnableIdCache)
	gamma_api.TableAddEnableRealtime(builder, table.EnableRealtime)
	if len(indexesOffsets) > 0 {
		gamma_api.TableAddIndexes(builder, indexesVec)
	}
	builder.Finish(builder.EndObject())
	return builder.FinishedBytes()
}

func (table *Table) DeSerialize(buffer []byte) {
	table.table = gamma_api.GetRootAsTable(buffer, 0)
	table.Name = string(table.table.Name())
	table.Fields = make([]FieldInfo, table.table.FieldsLength())
	for i := 0; i < len(table.Fields); i++ {
		var fieldInfo gamma_api.FieldInfo
		table.table.Fields(&fieldInfo, i)
		table.Fields[i].Name = string(fieldInfo.Name())
		table.Fields[i].DataType = DataType(fieldInfo.DataType())
		if fieldInfo.IndexType() != 0 {
			table.Fields[i].IndexType = fieldInfo.IndexType()
			table.Fields[i].IsIndex = table.Fields[i].IndexType != 0
		} else {
			table.Fields[i].IsIndex = fieldInfo.IsIndex()
			if table.Fields[i].IsIndex {
				table.Fields[i].IndexType = 1
			}
		}
	}

	table.VectorsInfos = make([]VectorInfo, table.table.VectorsInfoLength())
	for i := 0; i < len(table.VectorsInfos); i++ {
		var vecInfo gamma_api.VectorInfo
		table.table.VectorsInfo(&vecInfo, i)
		table.VectorsInfos[i].Name = string(vecInfo.Name())
		table.VectorsInfos[i].DataType = DataType(vecInfo.DataType())
		table.VectorsInfos[i].IsIndex = vecInfo.IsIndex()
		table.VectorsInfos[i].Dimension = vecInfo.Dimension()
		table.VectorsInfos[i].StoreType = string(vecInfo.StoreType())
		table.VectorsInfos[i].StoreParam = string(vecInfo.StoreParam())
	}

	table.IndexType = string(table.table.IndexType())
	table.IndexParams = string(table.table.IndexParams())
	table.RefreshInterval = table.table.RefreshInterval()
	table.EnableIdCache = table.table.EnableIdCache()
	table.EnableRealtime = table.table.EnableRealtime()

	table.Indexes = make([]IndexInfo, table.table.IndexesLength())
	for i := 0; i < len(table.Indexes); i++ {
		var idx gamma_api.IndexInfo
		table.table.Indexes(&idx, i)
		table.Indexes[i].Name = string(idx.Name())
		table.Indexes[i].Type = string(idx.Type())
		table.Indexes[i].FieldName = string(idx.FieldName())
		table.Indexes[i].FieldNames = make([]string, idx.FieldNamesLength())
		for j := 0; j < len(table.Indexes[i].FieldNames); j++ {
			table.Indexes[i].FieldNames[j] = string(idx.FieldNames(j))
		}
	}
}
